core/src/test/scala/org/apache/spark/storage/MemoryStoreSuite.scala

```diff
@@ -291,11 +290,11 @@ class MemoryStoreSuite
     blockInfoManager.removeBlock("b3")
     putIteratorAsBytes("b3", smallIterator, ClassTag.Any)
 
-    // Unroll huge block with not enough space. This should fail and kick out b2 in the process.
+    // Unroll huge block with not enough space.
     val result4 = putIteratorAsBytes("b4", bigIterator, ClassTag.Any)
     assert(result4.isLeft) // unroll was unsuccessful
     assert(!memoryStore.contains("b1"))
-    assert(!memoryStore.contains("b2"))
+    assert(memoryStore.contains("b2")) // not necessarily evicted
```

Comment:
This behavior changed after the change above. However, I am not sure this assertion is correct? there are 4800 bytes available. b2 takes 400. An attempt to put 4000 may fail (? well that was the previous and current behavior, anyway), but I don't see that it necessarily requires evicting b2; the allocation failed. We've recently had a very similar discussion about a different change that also uncovered this assertion as an issue.

Reply:
isn't b2 takes 4000 and b4 takes 40000 ?

I verified the test and I think the assertion is correct. 

Actually, for current behaviour with `UnifiedMemoryManager`, there may be 12000 bytes available. When we try to put b4 into memory, it will require 24064 memory once it unrolled 16 elements (which hit the `memoryCheckPeriod` for first time). And 24064 bytes exceed the maxMemory of 12000, so, it would just return instread of evicting any blocks.